clippy_lints/src/path_from_format.rs

```diff
@@ -0,0 +1,101 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+use std::fmt::Write as _;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `PathBuf::From(format!(..))` calls.
+    ///
+    /// ### Why is this bad?
+    /// It is not OS-agnostic, and can be harder to read.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::path::PathBuf;
+    /// let base_path = "/base";
+    /// PathBuf::from(format!("{}/foo/bar", base_path));
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::path::Path;
+    /// let base_path = "/base";
+    /// Path::new(base_path).join("foo").join("bar");
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub PATH_FROM_FORMAT,
+    pedantic,
+    "builds a `PathBuf` from a format macro"
+}
+
+declare_lint_pass!(PathFromFormat => [PATH_FROM_FORMAT]);
+
+impl<'tcx> LateLintPass<'tcx> for PathFromFormat {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if_chain! {
+            if let ExprKind::Call(_, args) = expr.kind;
+            if let ty = cx.typeck_results().expr_ty(expr);
+            if is_type_diagnostic_item(cx, ty, sym::PathBuf);
+            if !args.is_empty();
+            if let Some(macro_def_id) = args[0].span.ctxt().outer_expn_data().macro_def_id;
+            if cx.tcx.get_diagnostic_name(macro_def_id) == Some(sym::format_macro);
+            then {
+                let full_expr = snippet(cx, expr.span, "error").to_string();
+                let split_expr: Vec<&str> = full_expr.split('!').collect();
+                let args_to_macro = split_expr[1];
+                let replaced = args_to_macro.replace('(', "").replace(')', "");
+                let unformatted: Vec<&str> = replaced.split(',').collect();
+                let mut push_targets: Vec<String> = Vec::new();
+                let mut temp_string = String::new();
+                for c in unformatted[0].chars() {
+                    if c == '/' || c == '\\' {
+                        push_targets.push(temp_string.clone());
+                        temp_string = String::new();
+                    }
+                    else if c == '}' {
+                        temp_string.push_str(&unformatted[1].replace(' ', ""));
+                    }
+                    else if c != '{' && c != '"' {
+                        temp_string.push(c);
+                    }
+                }
```

Comment:
Instead of doing it manually, there's a clippy util to parse `format!` arguments from `Expr`s, maybe we can use that instead 

https://github.com/rust-lang/rust-clippy/blob/f0bf2003de1e50bc722941dc07b6ca593a136000/clippy_utils/src/macros.rs#L449-L467